scio-extra/src/main/scala/com/spotify/scio/extra/json/nio/JsonIO.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2018 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.extra.json.nio
+
+import com.spotify.scio.ScioContext
+import com.spotify.scio.extra.json.DecodeError
+import com.spotify.scio.io.Tap
+import com.spotify.scio.nio.{ScioIO, TextIO}
+import com.spotify.scio.util.ScioUtil
+import com.spotify.scio.values.SCollection
+import io.circe.{Decoder, Encoder, Printer}
+import io.circe.parser._
+import io.circe.syntax._
+import org.apache.beam.sdk.io.{Compression, FileBasedSink, TextIO => BTextIO}
+
+import scala.concurrent.Future
+import scala.reflect.ClassTag
+import scala.util.{Left, Right}
+
+case class JsonIO[T: ClassTag : Encoder : Decoder](path: String)
+  extends ScioIO[Either[DecodeError, T]] {
```

Comment:
would it be worth it to consider other `ADT` instead of plain `Either` to give better semantic? i.e:

```scala
sealed trait JsonSerDe[+T]
final case class Error(error: io.circe.Error, input: String) extends JsonSerDe[Nothing]
final case class Elem[T](e: T) extends JsonSerDe[T]
```